### PR TITLE
Moved unique_lock of messages_mutex_ to guarantee pointer

### PR DESCRIPTION
**Pull request by marting87**
**messages_mutex_** will be locked after **msg_it** was set. When publishing a laser scan at a high rate (150Hz) or when when you have any hick-ups in your pipeline it is not guaranteed that the iterator msg_it is still pointing at the expected element in messages_.

This leads to memory corruption and a crashing amcl. 

Corrected message_mutex to be in line with original (ROS1) code:
https://github.com/ros/geometry2/blob/noetic-devel/tf2_ros/include/tf2_ros/message_filter.h

Related: https://github.com/ros2/geometry2/issues/213
Possibly related: https://github.com/ros2/geometry2/issues/200

### CODE REVIEW

**Comment by allenh1** [comment]
That definitely seems possible to me. My only concern is how this patch might affect performance. Do you have any round trip comparisons before and after this is applied?

If the hit is too large maybe we could consider a slightly different approach that doesn't rely on iterators?

**Comment by clalancette** [comment]
> That definitely seems possible to me. My only concern is how this patch might affect performance. Do you have any round trip comparisons before and after this is applied?

That's one concern, though honestly that one is somewhat secondary to me.  The bigger concern is that `canTransform` can block for arbitrary amounts of time; since this patch now takes the mutex *before* calling `canTransform`, it means that it can now starve out all other users on other threads.  I don't think that is desirable.

However, I do have a thought on how to fix this up so it doesn't block in `canTransform` but still adds additional thread safety.  I'll add comments inline.

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 1]
```diff
@@ -429,7 +429,8 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
         --message_count_;
       }
 
-      // Add the message to our list
+      // Add the message to our list while keeping a lock on the messages.
+      std::unique_lock<std::mutex> unique_lock(messages_mutex_);
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/message_filter.h]: Instead of adding this here, just move this to the beginning of the else block (https://github.com/ros2/geometry2/pull/279/files#diff-724ff79b0edd5a83f8a62b5e2ffa085cR412), then https://github.com/ros2/geometry2/pull/279/files#diff-724ff79b0edd5a83f8a62b5e2ffa085cR416 .  That should cover both cases.

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 2]
```diff
@@ -491,6 +492,9 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
   {
     namespace mt = message_filters::message_traits;
 
+    // We will be accessing and mutating messages now, require unique lock
+    std::unique_lock<std::mutex> lock(messages_mutex_);
+
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/message_filter.h]: As I pointed out in the comment, this isn't desirable since it holds onto the lock the entire time, even over `canTransform`.  However, if we look carefully at what this code does, it does looking for the correct item in the list, and assuming it finds it, it does an operation on it and removes it from the list.  So what I'll suggest we do instead here is something like the following pseudo-code:

```
{
    lock;
    for (; msg_it != msg_end; ++msg_it) {
        handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
        if (handle_it != info.handles.end()) {
            info.success_count++;
            if (info.success_count >= expected_success_count_) {
              *saved_msg = msg_it;
              messages_.erase(msg_it);
              break;
            }
        }
    }
}
```

The idea here is to do all of the lookup and erasure of the item in the list under the lock at the beginning.  Then we drop the lock, and the rest of the code below operates on the thing we've removed from the list.  This should improve thread safety, but still not hold the lock across the `canTransform` calls.

Let me know if you think this will work, or don't understand what I am talking about here.

**Review by clalancette — CHANGES_REQUESTED** [review]

**Comment by allenh1** [comment]
> The bigger concern is that canTransform can block for arbitrary amounts of time; since this patch now takes the mutex before calling canTransform

That's a very good point.

I like the solution you proposed, looks quite reasonable to me.

File: tf2_ros/include/tf2_ros/message_filter.h (hunk 2 above)
Comment (marting87) [reply tf2_ros/include/tf2_ros/message_filter.h]: Good point! I've made the adjustment, however this does require an additional copy of messages. ```saved_msg = *msg_it;```

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 3]
```diff
@@ -494,27 +495,37 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
...
+        auto handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
+        if (handle_it != info.handles.end()) {
+          // found msg_it
+          ++info.success_count;
+          if (info.success_count >= expected_success_count_) {
+            saved_msg = *msg_it;
+            messages_.erase(msg_it);
+            --message_count_;
+          } else {
+            return;
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/message_filter.h]: From a purely stylistic perspective, I find this hard to follow.  Instead, I'll suggest we do:

```
msg_it = msg_end;
```

And then let it fall through to hit the break, which will then cause it to return in the conditional below.

Alternatively, we could move the break inside of the `if` brackets, which would also be slightly easier to reason about in my opinion.

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 4]
```diff
@@ -494,27 +495,37 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
...
+      std::unique_lock<std::mutex> lock(messages_mutex_);
+
+      for (; msg_it != msg_end; ++msg_it) {
+        MessageInfo & info = *msg_it;
+        auto handle_it = std::find(info.handles.begin(), info.handles.end(), handle);
+        if (handle_it != info.handles.end()) {
+          // found msg_it
+          ++info.success_count;
+          if (info.success_count >= expected_success_count_) {
+            saved_msg = *msg_it;
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/message_filter.h]: Now that I look at this again, it looks like once we are out of that initial loop, we only ever use the "event" from the `MessageInfo` struct.  So we can save some copying here by changing `saved_msg` to an `MEvent`, and just doing:

```
saved_event = msg_it->event;
```

File: tf2_ros/include/tf2_ros/message_filter.h [hunk 5]
```diff
@@ -753,3 +759,4 @@ class MessageFilter : public MessageFilterBase, public message_filters::SimpleFi
 } // namespace tf2
 
 #endif
+
```
Comment (clalancette) [inline tf2_ros/include/tf2_ros/message_filter.h]: Nit: please don't add trailing whitespace here.

**Review by clalancette — CHANGES_REQUESTED** [review]
> Good point! I've made the adjustment, however this does require an additional copy of messages.

Yeah, I figured that was the case.  See inline for an idea which saves a bit of copying.  There are also a couple of other nits inline.

Overall though, looking good!

**Review by clalancette — APPROVED** [review]
This one looks good to me!  Thanks for iterating.  I'll run CI on it now.

**Comment by clalancette** [comment]
CI:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/11290/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6537/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/9242/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/11193/)

**Comment by clalancette** [comment]
On macOS and Windows, there are some test failures.

macOS has 6 failures:

* projectroot.projection_test - flakey, also failed in https://ci.ros2.org/view/nightly/job/nightly_osx_debug/1673, so we can ignore
* projectroot.test.test_publisher_subscription_count_api - fails every night, so we can ignore
* projectroot.test.test_subscription_publisher_count_api - fails every night, so we can ignore
* projectroot.test_static_publisher.launch.py - flakey, also failed in https://ci.ros2.org/view/nightly/job/nightly_osx_release/1700, so we can ignore
* laser_geometry.projection_test.gtest.missing_result - flakey, also failed in https://ci.ros2.org/view/nightly/job/nightly_osx_debug/1673, so we can ignore
* test_tf2.BufferClientTestAfterShutdown.test_exit_code - flakey, also failed in https://ci.ros2.org/view/nightly/job/nightly_osx_release/1700, so we can ignore

Windows has 34 failures:
* projectroot.src.security.builtin_plugins.tests.CUnit* tests - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1594, so we can ignore
* projectroot.test_rate - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1591, so we can ignore
* projectroot.rqt_py_common - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1593, so we can ignore
* projectroot.gtest_timeout_subscriber__rmw_connext_cpp - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1593, so we can ignore
* test_rclcpp.test_timeout_subscriber__rmw_connext_cpp.timeout_subscriber - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1593, so we can ignore
* ros2bag.test.test_record_qos_profiles.test_record_qos_profiles - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1594, so we can ignore
* rclpy.src.ros2.rclpy.rclpy.test.test_rate.TestRate.test_rate_valid_period - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1591, so we can ignore
* rqt_py_common.rqt_py_common.xunit.missing_result - also failed in https://ci.ros2.org/view/nightly/job/nightly_win_rel/1593, so we can ignore

With that, this is good to go.  I'm going to go ahead and merge.  Thank you for the fix and for iterating here.

**Comment by marting87** [comment]
You're welcome @clalancette ! Is it possible to backport this fix to eloquent?

**Comment by clalancette** [comment]
> You're welcome @clalancette ! Is it possible to backport this fix to eloquent?

Yeah, we should also do Foxy as well.  Would you mind opening new PRs targeting those branches?  Make sure to put "Eloquent backport" and "Foxy backport" in the title.  Thanks.
